### ISSUE TITLE
[Bug]: PC0038 - False Positive with named return value and error

### ISSUE THREAD

**Issue by pri-kise** [ALCops/Analyzers#473]
### Analyzer

ApplicationCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0038

### Description

I receive the info:
> 'GetJsonToken(JsonObject, Text)': not all code paths return a value

I think that this is wrong.

### AL code to reproduce

```al
    internal procedure GetJsonToken(JObject: JsonObject; TokenKey: Text) JToken: JsonToken
    var
        CouldNotFindTokenErr: Label 'Could not find a token with key %1!', Comment = '%1 = TokenKey';
    begin
        if not JObject.Get(TokenKey, JToken) then
            Error(CouldNotFindTokenErr, TokenKey);
    end;
```

**Comment by MODUSCarstenScholling** [comment]
I'll catch this in the next week.

**Comment by pibcht** [comment]
Hi, I have the same issue 

Another example from my codebase :

```al
    internal procedure GetAPIAuthenticationToken() Value: SecretText
    begin
        if not this.TryGetIsolatedStorage(Value, APIAuthenticationToken) then
            Clear(APIAuthenticationToken);
    end;
```

And also:

```al
    internal procedure Get(TextEncoding: TextEncoding) Encoding: Enum BlobEncoding
    begin
        case TextEncoding of
            TextEncoding::MSDos:
                Encoding := Encoding::MSDos;
            TextEncoding::UTF16:
                Encoding := Encoding::UTF16;
            TextEncoding::Windows:
                Encoding := Encoding::Windows;
            TextEncoding::UTF8:
                Encoding := Encoding::UTF8;
        end;
    end;

    internal procedure Get(Encoding: Enum BlobEncoding) TextEncoding: TextEncoding
    begin
        case Encoding of
            Encoding::MSDos:
                TextEncoding := TextEncoding::MSDos;
            Encoding::UTF16:
                TextEncoding := TextEncoding::UTF16;
            Encoding::Windows:
                TextEncoding := TextEncoding::Windows;
            Encoding::UTF8:
                TextEncoding := TextEncoding::UTF8;
        end;
    end;
```

**Comment by MODUSCarstenScholling** [comment]
Included in #505

**Comment by stephanweidmann** [comment]
is the case mentioned in https://github.com/ALCops/Analyzers/discussions/510 also covered by #505 ?

**Comment by MODUSCarstenScholling** [comment]
> is the case mentioned in [#510](https://github.com/ALCops/Analyzers/discussions/510) also covered by [#505](https://github.com/ALCops/Analyzers/pull/505) ?

To be honest, whoever thought named return values were a great idea clearly didn’t ask me first. 🙂

I understand the use case. I could add specific workarounds for records and other complex datatypes, but they would only cover a limited set of scenarios and would continue to raise false positives, since custom procedures may populate the complex datatype as well.

Ultimately, this and similar constructs cannot work reliably without an explicit exit and moving the named return value to a local variable.

So, my suggestion is: Consider using an explicit `exit()` or use `#pragma warning disable PC0036`in these cases.

**Comment by MODUSCarstenScholling** [comment]
Oh my gosh, I sound just like a politician. The answer to your question is “no.” :)

**Comment by stephanweidmann** [comment]
thank you, have a nice weekend
